Make install hooks viral

azdavis opened the pull request:
## Motivation

This PR is pre-work for the final runtime checks. It also contains some other
assorted minor fixes.

When final things (methods and modules) are implemented in the runtime, we would
like for all consumers of final things to have all of the final checks be run
for them, at all times. This means that even if a consumer of a final thing
isn't using other features of the Sorbet runtime like `sig`s in their own code,
we would still like to make sure they are respecting the final checks.

This means that, for instance, if a module `Consumer` includes a module
`HasFinal`, where `HasFinal` defines some final methods, we would like all
further method definitions in `Consumer` to make sure that they are not
overriding a final method definition in `HasFinal`.

The way that we listen in for and do things with method definitions is the
runtime hooks. Whenever a module utters `extend T::Sig` followed by a call to
`sig`, that module has the hooks installed on itself. These hooks then listen in
for new method definitions on that module, and wrap those method definitions
with the preceding `sig` logic, if there was any.

Before this PR, the only way the hooks could be installed on a module (other
than explicitly calling `T::Private::Methods::install_hooks(mod)`) was by
uttering `sig`. However, as previously mentioned, we would like for the final
checks to be run for consumers of final methods, even if a consumer doesn't use
`sig`.

Because I plan to implement the final checks using the existing hooks machinery
(i.e. by adding to `T::Private::Methods._on_method_added`), this means that we
would like to install the hooks on any module `Consumer` that includes or
extends a module `HasFinal`, where `HasFinal` contains some final methods.

This PR implements approximately that, in that, we now install the hooks on any
module `Consumer` that includes or extends a module `HasHooks`, where `HasHooks`
is a module that has the hooks installed on it. This works because the only way
to mark a method as final is to say so in its `sig`, and once you have put a
`sig` on a method, you have the hooks installed on yourself.

This means that when final modules are added, we will probably want to install
the hooks on any module that utters `final!` in its body.

The way we implement the above logic is by tapping into the `included` and
`extended` methods in Ruby, which are called when a module is included or
extended. When a module has the hooks installed on itself, we add a bit of logic
to its `included` and `extended` that make it such that the includer or extender
also gets the hooks.

## Pitfalls (OUTDATED)

Reviewers of this PR who review by commit will notice that the first pass at the
implementation looked like this:

```rb
orig_included = T::Private::ClassUtils.replace_method(mod.singleton_class, :included) do |arg|
  # ... our logic to apply the hooks to arg ...
  orig_included.bind(self).call(arg)
end
```

Here, `orig_included` is the `included` that was defined on
`mod.singleton_class` at the time that the call to
`T::Private::ClassUtils.replace_method` was made. The problem with this is that,
if, after this call to `replace_method`, we define a different `included` on
`mod.singleton_class`, that `orig_included` does not refer to it.

This was exemplified by modules that `include T::Props::Plugin` not behaving
correctly.

The fix for this bug is to change the implementation to this:

```rb
T::Private::ClassUtils.replace_method(mod.singleton_class, :included) do |arg|
  # ... our logic to apply the hooks to arg ...
  super(arg)
end
```

Where `super` (appears to) simply refer to "the included before us". (I'm not
100% on the semantics here, but all tests pass so I think this is right.)

The reason why I'm specifically calling this out here is because there seems to
be lots of places in the sorbet runtime where we use the general pattern shown
in the first example, and I wonder if any of them are subtly incorrect the way
my first implementation was.

## A third pass

The above section has been marked outdated, but has been kept for posterity.

A third pass at the implementation avoids `replace_method` altogether. We
instead defines a `module VirallyInstallHooks` which defines the `included` and
`extended` we want, together with the appropriate `super` to make sure we still
run the other `included` and `extended`. We then have modules extend this
`VirallyInstallHooks`.

## Test plan

Existing tests pass. New tests added.

jez commented:
> The reason why I'm specifically calling this out here is because there seems to
> be lots of places in the sorbet runtime where we use the general pattern shown
> in the first example, and I wonder if any of them are subtly incorrect the way
> my first implementation was.

@azdavis can you test this on pay-server?

jez commented:
> Where `super` (appears to) simply refer to "the included before us". (I'm not
> 100% on the semantics here, but all tests pass so I think this is right.)

@DarkDimius do you know if the distinction between `orig_included.bind(self).call(...)` on the original `included` method vs `super(...)` is significant? Could this cause problems?

jez commented:
@azdavis is there really no test you can write to verify whether the new things in this PR are correct? Something that evidences the virality of the hooks?

azdavis commented:
The tests in the branch for the PR that implements final method check now pass when rebased onto this, so I think they're working. I will add some more explicit tests in this PR too.

azdavis commented:
> @azdavis can you test this on pay-server?

What do you mean by this?

azdavis commented:
In the latest version of this PR, I changed the following:

First, in order to have my logic in `included` and `extended` be run while
preserving previous and future `included` and `extended`, I defined a
`module VirallyInstallHooks` which I then extended as appropriate. See the
updated PR description and the commits.

Second: The criteria for virally spreading hooks to a module A which includes or
extends a module B used to be "B has the hooks". The criteria is now "B is in
the specific set of modules for which we want viral spreading of hooks to
occur." This set (`@virally_install_hooks` on `T::Private::Methods`) shall be
populated as necessary once the final checks start rolling in, but is currently
empty. This change means that, as a direct result of this PR, no modules start
getting the hooks when previously they did not. I made this change to fix some
errors on pay-server.

This change seems to make pay-server pass when it did not before. That would
seem to imply that installing the hooks on certain modules in pay-server makes
things break in unexpected ways. However, instead of investigating for exactly
what modules this is the case, and why this is the case, I opted to just be more
conservative with who to spread the hooks to. We may want to investigate this.

jez commented on gems/sorbet-runtime/lib/types/private/methods/_methods.rb:
```diff
@@ -300,10 +298,36 @@ def self.run_all_sig_blocks
+  #   then we want B to have the hooks installed on itself
+  # should extend this module.
+  #
+  # note that the arg is by default nil so it doesn't conflict with `Chalk::Tools::Concern` in pay-server.
```
This seems like a gotcha. What do you mean "conflict"? The correctness of sorbet-runtime shouldn't depend on how pay-server or any other ruby codebase works.

azdavis replied:
Actually, this should no longer be necessary because we're not adding new logic for `included` and `extended` for any modules. But I was running into issues earlier with the number of arguments `include` being optionally 0 if you `extend Chalk::Tools::Concern`, which defines its own DSL with a method called `included` that takes as few as 0 arguments but also takes a block.

jez commented on gems/sorbet-runtime/test/types/virally_install_hooks.rb:
```diff
@@ -0,0 +1,58 @@
+
+  it "gives the hooks when including a viral module" do
+    m = Module.new do
+      T::Private::Methods.instance_variable_get(:@virally_install_hooks) << self
```
Is the point of this instance variable just to be able to write this test?

azdavis replied:
Hm. The point of this instance variable, in my mind, was to make it so that a module which, for example, defines a final method, would get the new `included` and `extended` behavior.

However, one thing I failed to account for is that a module might not have any final methods on it at the time it has `install_hooks` called on it, but it could gain final methods later. I think that means instead of using this `@virally_install_hooks`, it might be better to do something like just call `mod.extend(VirallyInstallHooks)` whenever a final method is added to `mod`. (It appears to be safe to call `extend` multiple times.)

In a related vein, I'm wondering if it was a mistake to try to split this PR away from the final method checks. You left another comments about the tests I wrote, but I wonder if it would be better to test if things are working by seeing if the final checks are running when we want them to be. WDYT about closing this PR and folding it into #1181?

jez replied:
I think it's fine to fold them into one if you want.

jez commented on gems/sorbet-runtime/test/types/virally_install_hooks.rb:
```diff
@@ -0,0 +1,58 @@
+    end
+    assert_equal(true, T::Private::Methods.instance_variable_get(:@installed_hooks).include?(c))
+  end
+end
```
It seems like the primary property we're trying to test is not the one being tested here. Our use of `instance_variable_get` makes me think that this test is really just testing that instance_variable_set / instance_variable_get still work in Ruby. The thing we really want to know is if we do this:

```
include m1
include m2
```

that our method added hooks get run for any methods defined in `m2`, right? Instead of reaching into instance variables, can we test that property directly?

azdavis replied:
Addressed above.

jez commented on gems/sorbet-runtime/test/types/virally_install_hooks.rb:
```diff
@@ -0,0 +1,58 @@
+    assert_equal(true, T::Private::Methods.instance_variable_get(:@installed_hooks).include?(m))
+    c = Class.new do
+      extend m
+    end
```
If one of these Classes has

```
def included; end
def extended; end
```

i.e., does not call `super`, does that prevent the things you introduced in this branch from working?

azdavis replied:
Almost certainly. Should I test for this explicitly? Should we try to prevent this?

jez replied:
It would seem that we definitely want to prevent any way to change a final method.

azdavis commented:
Per [discussion][1], closing this in favor of #1181.

[1]: https://github.com/sorbet/sorbet/pull/1241#discussion_r302297548
